README.md
# Send SMS
Helper class to work with Mauritius operator SMS gateway. Note that you need credentials to be able to use this package.

## Features
- Send SMS using operator API
- Get transaction reference, error code and description as response

## Installation

```bash
$ composer require aliirfaan/mu-sms-kit
```

## Basic Usage

```php
<?php

require 'vendor/autoload.php';

use  Aliirfaan\SmsKit\Sms;

// instantiate class with your credentials
$sms = new Sms('my_username', 'my_password', 'my_sender_id', 'my_api_endpoint');

// use
$message = 'Hello';
$phoneNumber = '23057777777';

$sendSms = $sms->sendSms($phoneNumber, $message);
```

### Requirements
- PHP 5.6.0 or above
- cURL

## License

The MIT License (MIT)

Copyright (c) 2020

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com) and this project adheres to [Semantic Versioning](https://semver.org).

## 2.1.1 - 2021-07-12

### Added

- Nothing

### Changed

- getTransactionReference() cater for responses without dash '-'

### Deprecated

- Nothing

### Removed

- Nothing

### Fixed

- Nothing

## 2.1.0 - 2021-02-18

### Added

- Nothing

### Changed

- smsWasSent() add new parameter $validTransactionReferenceMinLength

### Deprecated

- Nothing

### Removed

- Nothing

### Fixed

- Nothing

## 2.0.1 - 2021-02-18

### Added

- Nothing

### Changed

- sendSms() return transaction reference in message

### Deprecated

- Nothing

### Removed

- Nothing

### Fixed

- Nothing

## 2.0.0 - 2021-02-18

### Added

- Nothing

### Changed

- curl request
- sendSms() added $curlOptions as parameter so that you can pass your own cURL options

### Deprecated

- Nothing

### Removed

- params key in response
- remove CURLOPT_NOBODY, CURLOPT_HEADER in default cURL options

### Fixed

- Nothing

tests/SmsTest.php
<?php
namespace Aliirfaan\SmsKit\Tests;

use PHPUnit\Framework\TestCase;
use Aliirfaan\SmsKit\Sms;

/**
*  Test for myt-sms helper class
*  @author aliirfaan
*/
class SmsTest extends TestCase
{
    private $username = 'test';

    private $password = 'test';

    private $senderId = 'test';

    private $apiUrl = 'test';

    public function providerTestSmsWasSent()
    {
        return array(
            array('0306201020101331756371100', true),
            array('91', false),
            array('4000', false),
            array('01', false),
            array('7', false),
            array('', false)
        );
    }

    public function providerTestGetTransactionReference()
    {
        return array(
            array('23057723610-0306201020101331756371100', '0306201020101331756371100'),
            array('7713611-91', '91'),
            array('7713611-01', '01'),
            array('7713611-', ''),
            array('91', ''),
            array('9991', ''),
            array('abc', ''),
            array('', '')
        );
    }

    /**
     * @dataProvider providerTestSmsWasSent
     */
    public function testSmsWasSent($transactionReference, $expectedResult)
    {
        $sms = new Sms($this->username, $this->password, $this->senderId, $this->apiUrl);

        $result = $sms->smsWasSent($transactionReference);
        $this->assertEquals($expectedResult, $result['success']);
    }

    /**
     * @dataProvider providerTestGetTransactionReference
     */
    public function testGetTransactionReference($response, $expectedResult)
    {
        $sms = new Sms($this->username, $this->password, $this->senderId, $this->apiUrl);

        $result = $sms->getTransactionReference($response);
        $this->assertEquals($expectedResult, $result);
    }
}


src/GatewayResponse.php
<?php
namespace Aliirfaan\SmsKit;

/**
 * GatewayResponse
 * @author aliirfaan
 */
class GatewayResponse
{
    /**
     * Status codes translation table.
     *
     * The list of codes returned by SMS API
     *
     * @var array
     */
    public const statusTexts = [
        '4000' => 'MAX_QUERY_STRING_LENGTH',
        '100' => 'MAX_PARAM_NAME_LENGTH',
        '3000' => 'MAX_PARAM_VALUE_LENGTH',
        '14' => 'MAX_HTTP_ORIG_ADDRESS_LEN',
        '16' => 'HTTP_MAX_MOBILENUMBER_LEN',
        '7' => 'MAX_HTTP_CLIENTID_LEN',
        '134' => 'HTTP_INVALID_CONTRY_CODE',
        '133' => 'HTTP_FLASH_MSG_LENGTH_EXCEEDED',
        '132' => 'HTTP_MESSAGE_LIMIT_EXCEEDED',
        '131' => 'HTTP_DBINSERT_FAILED',
        '130' => 'HTTP_HOLIDAYRESTRICTION_FAILED',
        '129' => 'HTTP_INVALID_PROTOCOL',
        '128' => 'HTTP_DBINIT_FAILED',
        '127' => 'HTTP_HTML_DIR_NOTSET',
        '126' => 'HTTP_CREATE_TARFILE_ERROR',
        '125' => 'HTTP_DATA_NOTAVAILABLE',
        '124' => 'HTTP_FILEOPEN_ERROR',
        '123' => 'HTTP_FILE_MISSING',
        '122' => 'HTTP_FILEOPEN_FAILED',
        '121' => 'HTTP_INVALID_FILE',
        '120' => 'HTTP_TIME_EXCEDED',
        '119' => 'HTTP_TIME_ELAPSED',
        '118' => 'HTTP_WRONGDATEFORMAT',
        '117' => 'HTTP_DATEFORMAT_MISSING',
        '116' => 'HTTP_SECOND_MISSING',
        '115' => 'HTTP_MINUTE_MISSING',
        '114' => 'HTTP_HOUR_MISSING',
        '113' => 'HTTP_DATE_MISSING',
        '112' => 'HTTP_CONCAT_MSG_MISSING',
        '111' => 'CFGPATH_NOTSET',
        '110' => 'SCRIPT_FILENAME_NOTSET',
        '109' => 'HOMEPATH_NOTSET',
        '108' => 'HTTP_INVALIDTIMEZONE_FAILED',
        '107' => 'HTTP_LOGINIT_FAILED',
        '106' => 'HTTP_GETCLIENTINFO_FAILED',
        '105' => 'HTTP_ACNTACTIVE_FAILED',
        '104' => 'HTTP_LOGINCLIENT_FAILED',
        '103' => 'HTTP_LMTCTRL_INT_FAILED',
        '102' => 'HTTP_MSGCTRL_INIT_FAILED',
        '101' => 'HTTP_WRONG_CLIENTID',
        '100' => 'HTTP_MSGCRTL_FAILED',
        '100' => 'HTTP_THROUPUT_NOT_AVAILABLE',
        '99' => 'HTTP_USERNAME_MISSING',
        '98' => 'HTTP_PASSWORD_MISSING',
        '97' => 'HTTP_MOBILE_NUMBER_MISSING',
        '96' => 'HTTP_USERDATA_MISSING',
        '95' => 'HTTP_DB_FAILURE',
        '94' => 'HTTP_SYSTEM_ERROR',
        '93' => 'HTTP_WRONG_UNAME',
        '92' => 'HTTP_WRONG_PASSWD',
        '91' => 'HTTP_INVALID_MOBILE_NUMBER',
        '90' => 'HTTP_UDH_MISSING',
        '01' => 'HTTP_DATA_AVAILABLE'
    ];
}


src/Sms.php
<?php
namespace Aliirfaan\SmsKit;

use Aliirfaan\SmsKit\GatewayResponse;

/**
 *  Use Messaging Gateway Platform to send SMS messages
 *  @author aliirfaan
 */
class Sms
{
    /**
     * The service endpoint
     *
     * @var string
     */
    private $apiUrl;
        
    /**
     * The account username
     *
     * @var string
     */
    private $username;

    /**
     * The account password
     *
     * @var string
     */
    private $password;
        
    /**
     * The account from sender Id
     *
     * @var string
     */
    private $senderId;

    /**
     * Create a new instance
     *
     * @param  string  $username
     * @param  string  $password
     * @param  string  $senderId
     * @param  string  $apiUrl
     *
     * @return void
     */
    public function __construct($username, $password, $senderId, $apiUrl)
    {
        $this->username = $username;
        $this->password = $password;
        $this->senderId = $senderId;
        $this->apiUrl = $apiUrl;
    }
    
    /**
     * Send a text message
     *
     * Use cURL to send request to a URL
     *
     * @param  string $phoneNumber
     * @param  string $message
     * @param  int $concatenated Possible values are true or false. If true the message is sent as concatenated if udata message is more than 160 chars.
     * @param  string $deferred
     * @param  string $dsr
     * @param  int $connectTimeoutSeconds Number of seconds that cURL should timeout
     * @param  array $curlOptions PHP supported cURL options https://www.php.net/manual/en/curl.constants.php Will take these options to setup cUrl if provided
     * @return array $data
     */
    public function sendSms($phoneNumber, $message, $concatenated = 1, $deferred = 'false', $dsr = 'true', $connectTimeoutSeconds = 60, $curlOptions = null)
    {
        $data = array(
            'result' => null,
            'success' => false,
            'message' => null,
        );
        
        $smsParams = array(
            'UserName' => $this->username,
            'PassWord' => $this->password,
            'SenderId' => $this->senderId,
            'Number' => $phoneNumber,
            'UserData' => $message,
            'Concatenated' => $concatenated,
            'Deferred' => $deferred,
            'Dsr' => $dsr
        );

        $requestParams = http_build_query($smsParams);
        $requestUrl = $this->apiUrl . '?' . $requestParams;

        try {
            $curl = curl_init();

            $curlSetupOptions = array(
                CURLOPT_RETURNTRANSFER => 1,
                CURLOPT_FAILONERROR  => 1,
                CURLOPT_CONNECTTIMEOUT  => $connectTimeoutSeconds,
                CURLOPT_URL => $requestUrl
            );

            if (!is_null($curlOptions)) {
                $curlSetupOptions = $curlOptions;
            }

            curl_setopt_array(
                $curl,
                $curlSetupOptions
            );
            
            $result = curl_exec($curl);
            if (!$result) {
                $data['message'] = 'Error: ' . curl_error($curl) . ' - Code: ' . curl_errno($curl);
            } else {
                $curlInfo = curl_getinfo($curl);
                $httpStatusCode = intval($curlInfo['http_code']);
                if ($httpStatusCode !== 200) {
                    $data['message'] = 'SMS was not sent. HTTP error code: '.$httpStatusCode;
                } else {
                    $result = trim($result);

                    $transactionReference = $this->getTransactionReference($result);
                    $smsWasSent = $this->smsWasSent($transactionReference);

                    $data['result'] = $result;
                    $data['success'] =  $smsWasSent['success'];
                    $data['message'] = $smsWasSent['message'];
                }
            }
            curl_close($curl);
        } catch (\Exception $e) {
            $data['message'] = 'An exception occured: '.$e->getMessage();
        }

        return $data;
    }
        
    /**
     * getTransactionReference
     *
     * Response is usually in this format: 23057713610-0306201020101331756371100
     * We explode at '-' to get transaction reference
     * We may also get response without '-'. Example: 91, 106
     *
     * @param  string $response
     * @return string transaction reference
     */
    public function getTransactionReference($response)
    {
        $transactionReference = $response;

        $responsePieces = explode('-', $response);
        if (\array_key_exists(1, $responsePieces)) {
            $transactionReference = $responsePieces[1];
        }
        $transactionReference = strval($transactionReference);

        return $transactionReference;
    }
    
    /**
     * smsWasSent
     *
     * Read transaction reference to see if sms was actually sent or we got an error code
     * Check if transaction reference exists in error code list
     * Check if transaction reference is blank
     * Check if transaction reference satifies a minimum length as successful transaction reference is long. Example: 0306201020101331756371100
     * 
     * @param  string $transactionReference
     * @param  int $validTransactionReferenceMinLength Minimum lenght of a valid transaction reference number
     * @return array
     */
    public function smsWasSent($transactionReference, $validTransactionReferenceMinLength = 5)
    {
        $data = array(
            'success' => false,
            'message' => null,
        );

        if (\array_key_exists($transactionReference, GatewayResponse::statusTexts)) {
            $data['message'] = 'SMS was not sent. Code: '.$transactionReference.' Description: '.GatewayResponse::statusTexts[$transactionReference];
        } elseif ($transactionReference == '') {
            $data['message'] = 'SMS was not sent. An unknown error occured';
        } elseif (strlen($transactionReference) < $validTransactionReferenceMinLength) {
            $data['message'] = 'SMS was not sent. '.$transactionReference.' is not an error code and does not satisfy minimum length';
        } else {
            $data['message'] = $transactionReference;
            $data['success'] = true;
        }

        return $data;
    }
}
